src/Compilers/Core/Portable/Operations/ControlFlowGraphBuilder.cs

```diff
@@ -0,0 +1,555 @@
+                                                                isImplicit: true));
+
+                    _evalStack[i] = new FlowCapture(captureId, operation.Syntax, isInitialization: false, operation.Type, operation.ConstantValue);
+                }
+            }
+        }
+
+        public override IOperation VisitSimpleAssignment(ISimpleAssignmentOperation operation, object argument)
+        {
+            _evalStack.Push(Visit(operation.Target));
+            IOperation value = Visit(operation.Value);
+            return new SimpleAssignmentExpression(_evalStack.Pop(), operation.IsRef, value, null, operation.Syntax, operation.Type, operation.ConstantValue, operation.IsImplicit);
+        }
+
+        // PROTOTYPE(dataflow):
+        //public override IOperation VisitArrayElementReference(IArrayElementReferenceOperation operation, object argument)
+        //{
+        //    _evalStack.Push(Visit(operation.ArrayReference));
+        //    foreach (var index in operation.Indices)
+        //    return new ArrayElementReferenceExpression(Visit(operation.ArrayReference), VisitArray(operation.Indices), ((Operation)operation).SemanticModel, operation.Syntax, operation.Type, operation.ConstantValue, operation.IsImplicit);
+        //}
+
+        private static bool IsConditional(IBinaryOperation operation)
+        {
+            switch (operation.OperatorKind)
+            {
+                case BinaryOperatorKind.ConditionalOr:
+                case BinaryOperatorKind.ConditionalAnd:
+                    return true;
+            }
+
+            return false;
+        }
+
+        public override IOperation VisitBinaryOperator(IBinaryOperation operation, object argument)
+        {
+            if (IsConditional(operation))
+            {
+                return VisitBinaryConditionalOperator(operation, true);
+            }
+
+            return base.VisitBinaryOperator(operation, argument);
+        }
+
+        public override IOperation VisitUnaryOperator(IUnaryOperation operation, object argument)
+        {
+            // PROTOTYPE(dataflow): ensure we properly detect logical Not
+            if (operation.OperatorKind == UnaryOperatorKind.Not)
+            {
+                return VisitConditionalExpression(operation.Operand, sense: false);
+            }
+
+            return base.VisitUnaryOperator(operation, argument);
+        }
+
+        private IOperation VisitBinaryConditionalOperator(IBinaryOperation binOp, bool sense)
+        {
+            bool andOrSense = sense;
+
+            switch (binOp.OperatorKind)
+            {
+                case BinaryOperatorKind.ConditionalOr:
+                    Debug.Assert(binOp.LeftOperand.Type.SpecialType == SpecialType.System_Boolean);
+                    Debug.Assert(binOp.RightOperand.Type.SpecialType == SpecialType.System_Boolean);
+
+                    // Rewrite (a || b) as ~(~a && ~b)
```

Comment:
Possibly a stupid question, but why rewrite this?

Reply:
> Possibly a stupid question, but why rewrite this?

Short-circuiting binary operators imply a control flow. For example, if ```a``` in ```a || b``` is true, we jump over evaluation of ```b```. This should be explicitly reflected in the graph through links. See LogicalOrFlow_01 unit-test in IOperationTests_IBinaryOperatorExpression.cs.

---
In reply to: [162218816](https://github.com/dotnet/roslyn/pull/24263#discussion_r162218816) [](ancestors = 162218816)